### PR TITLE
Add shatterseek

### PR DESCRIPTION
**Pull request by yangyangclover**
### Purpose/implementation Section
analyze chromothripsis, using modified shatterseek code

#### What scientific question is your analysis addressing?
Chromothripsis events in PBTA data

#### What was your approach?
modified shatterseek code

#### What GitHub issue does your pull request address?
[393](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/393)

#### Which areas should receive a particularly close look?
Test and Filter The Combined Results
Call Linking Groups

#### Is there anything that you want to discuss further?
we will describe the Chromothripsis Distribution across samples and different tumor types；
we will perform simple SV (SVs that are not in Chromothripsis regions) signatures analysis

#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?
A table called "PBTA_chromothripsis_newlink_region.csv" will be saved in "analyses\sv-analysis\tables\sv-shatterseek" folder, which is about the chromothripsis regions were founded in our dataset.

#### What types of results are included (e.g., table, figure)?
A table called "PBTA_chromothripsis_newlink_region.csv" will be saved in "analyses\sv-analysis\tables\sv-shatterseek" folder, which is about the chromothripsis regions were founded in our dataset.
We will use it in the following study.

#### What is your summary of the results?
the chromothripsis regions were founded in our dataset

#### Reproducibility Checklist

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [x] This analysis has been added to continuous integration.

#### Documentation Checklist

- [x] This analysis module has a `README` and it is up to date.
- [x] This analysis is recorded in the table in `analyses/README.md` and the entry is up to date.
- [x] The analytical code is documented and contains comments.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -232,8 +232,8 @@ RUN R -e "install.packages('corrplot', dependencies = TRUE)"
 # Install for mutation signature analysis
 RUN R -e "BiocManager::install('ggbio')"
 
-#### Please install your dependencies here
-#### Add a comment to indicate what analysis it is required for
+# package required for shatterseek
+RUN R -e  “remotes::install_github('parklab/ShatterSeek'), dependencies = TRUE”
```
Comment (jaclyn-taroni) [inline Dockerfile]: I believe the CI error is because the `)` is in the wrong spot. We'll also want to tie this to a specific commit to the ShatterSeek repository.
```suggestion
RUN R -e  “remotes::install_github('parklab/ShatterSeek', ref = '83ab3effaf9589cc391ecc2ac45a6eaf578b5046', dependencies = TRUE)”
```

File: Dockerfile [hunk 2]
```diff
@@ -232,8 +232,8 @@ RUN R -e "install.packages('corrplot', dependencies = TRUE)"
 # Install for mutation signature analysis
 RUN R -e "BiocManager::install('ggbio')"
 
-#### Please install your dependencies here
-#### Add a comment to indicate what analysis it is required for
+# package required for shatterseek
+RUN R -e  “remotes::install_github('parklab/ShatterSeek', ref = '83ab3effaf9589cc391ecc2ac45a6eaf578b5046', dependencies = TRUE)”
```
Comment (jaclyn-taroni) [inline Dockerfile]: Let's try:

```suggestion
RUN R -e  "remotes::install_github('parklab/ShatterSeek', ref = '83ab3effaf9589cc391ecc2ac45a6eaf578b5046', dependencies = TRUE)"
```

File: Dockerfile (hunk 1 above)
Comment (yangyangclover) [reply Dockerfile]: I updated the Dockerfile as you advised, but it displayed the same error and failed again.  BTW, may I ask what does the `ref = '83ab3effaf9589cc391ecc2ac45a6eaf578b5046'` mean?

Comment (cgreene) [reply Dockerfile]: What source code editor are you using? I've not seen one before that uses the directional quotes (`“`) instead of the standard ones `"`.

Comment (yangyangclover) [reply Dockerfile]: > What source code editor are you using? I've not seen one before that uses the directional quotes (`“`) instead of the standard ones `"`.

haha, sorry about that, that quote` “` is a quote in the Chinese input method, I didn't notice it.

Comment (jaclyn-taroni) [reply Dockerfile]: `ref = '83ab3effaf9589cc391ecc2ac45a6eaf578b5046'` is how you tie the installation to a specific commit. `83ab3effaf9589cc391ecc2ac45a6eaf578b5046` is the most recent commit.

**Comment by jaclyn-taroni** [comment]
@yangyangclover I am going to make changes directly to your branch to see if I can get things sorted out.

**Comment by jaclyn-taroni** [comment]
Okay I think we've done it with https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/449/commits/d7cc2759c1b27b8d7c306ab2ef190b19b58de149! It is very close to what you had in https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/449/commits/a7e57b13152807c0972a63378f0eebaa1c589c54 but changed `"true"` to `'true'` and added a `)` at the end.

**Comment by yangyangclover** [comment]
> Okay I think we've done it with [d7cc275](https://github.com/AlexsLemonade/OpenPBTA-analysis/commit/d7cc2759c1b27b8d7c306ab2ef190b19b58de149)! It is very close to what you had in [a7e57b1](https://github.com/AlexsLemonade/OpenPBTA-analysis/commit/a7e57b13152807c0972a63378f0eebaa1c589c54) but changed `"true"` to `'true'` and added a `)` at the end.

Thanks @jaclyn-taroni ! I learned a lot.

**Comment by jaclyn-taroni** [comment]
@yangyangclover I have not looked into the error coming up in CI in any detail, but I notice it might be a case of a missing biospecimen ID.

It looks like the file you're adding here (`analyses/sv-analysis/02-shatterseek.R`) consumes the consensus SEG file that's tracked in the repository and the output of `01-process-sv-file.R`. The `01-process-sv-file.R` output is _not_ tracked in the repository and that step runs before this one in CI. When things run in CI, this branch is checked out by the machine, so it will have access to all files that are included in the repo including the consensus SEG file. The consensus SEG file will contain _all_ samples that were run through the consensus copy number calls module (this should be every WGS sample). The Manta file that `01-process-sv-file.R` uses in CI will only have a subset of samples included.  Quoting the [**Continuous Integration (CI)** section](https://github.com/AlexsLemonade/OpenPBTA-analysis#continuous-integration-ci) of our README below:

> We have put together data files specifically for the purpose of CI that contain all of the features of the full data files for only a small subset of samples. You can see how this was done by viewing [this module](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/create-subset-files). We use the subset files to cut down on the computational resources and time required for testing.

So if your code assumes all samples (`Kids_First_Biospecimen_ID`) are present in both the copy number and SV data, that would cause a failure in CI.

This may not be the issue, but I wanted to give you that background information in case it helped you debug.

**Comment by cgreene** [comment]
@yangyangclover : are you still debugging this? It looks like it's nearly all commented out.

**Comment by yangyangclover** [comment]
> @yangyangclover : are you still debugging this? It looks like it's nearly all commented out.

I am still working with it, I'm sorry that I was too busy to debug this PR in the last two days. Thanks for your tracking. 👍

**Comment by cgreene** [comment]
Ok - thanks for letting us know!

**Comment by jaclyn-taroni** [comment]
Hi @yangyangclover, looks like the last build error has to do with the formatting of `.circleci/config.yml`. I can push a fix now so you can pull the change and continue debugging.

**Comment by jaclyn-taroni** [comment]
Looks like the spacing change in https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/449/commits/326e6ee8526f22752774171640e0ad035f56998a worked

**Comment by jaclyn-taroni** [comment]
@yangyangclover we need to get an analysis of chromothripsis added and this seems like it is not at a point where it can be reviewed yet. Do you have an idea of when you will be able to debug the CI issues and have it ready for review? We can try and help with the CI issues, let us know!

**Comment by yangyangclover** [comment]
Hi @jaclyn-taroni , I found there is something wrong in the `proccess_sv_file`, most of the output file is empty. I can send you my local results first if you need it, so you can start with them.

**Comment by jaclyn-taroni** [comment]
@yangyangclover if I recall correctly, `01-process-sv-file.R` creates a TSV file for each individual biospecimen/sample. Can you send me an example file that you have locally?

**Comment by yangyangclover** [comment]
@jaclyn-taroni  Yes, of course. I can send you the example TSV file. 

[BS_0DKPGQWD_withoutYandM.txt](https://github.com/AlexsLemonade/OpenPBTA-analysis/files/4135190/BS_0DKPGQWD_withoutYandM.txt)
[BS_0DKPGQWD.txt](https://github.com/AlexsLemonade/OpenPBTA-analysis/files/4135191/BS_0DKPGQWD.txt)
Since I couldn't upload `tsv` file, I changed the suffix to `txt`.

[GENES.docx](https://github.com/AlexsLemonade/OpenPBTA-analysis/files/4135192/GENES.docx)

Additionally, in case it takes too long to fix codes, do you need the shatterseek result so you can start analyzing now?

**Comment by jaclyn-taroni** [comment]
Thanks @yangyangclover! I want to make sure I fully understand - when I read in the file `BS_0DKPGQWD.txt`, there are 48 rows, which I understand to be 48 events. When I read in the SV file included as part of the data download (`pbta-sv-manta.tsv.gz`) and filter it to just the rows where `Kids.First.Biospecimen.ID.Tumor == "BS_0DKPGQWD"` and look at the SV types:

```r
BND DEL DUP INS INV 
226  39  23   1  18 
```
And we'd expect the `DEL`, `DUP`, and `INV` to each be represented + the (number of `BND`/2) because those become a single `TRA` in the output of `01-process-sv-file.R`. Do I have all of that correct?

**Comment by yangyangclover** [comment]
@jaclyn-taroni  correct!

**Comment by jaclyn-taroni** [comment]
Okay, great! I'll start a new branch and see what I can figure out!

**Comment by yangyangclover** [comment]
@jaclyn-taroni  Thank you so much!

**Comment by jaclyn-taroni** [comment]
@yangyangclover I think it's because of the filtering to `PASS`: https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/86bd634a928db3f460328210e38c7dc1797d1304/analyses/sv-analysis/01-process-sv-file.R#L48

48 seems like the correct number of events when I included that filtering.

**Comment by yangyangclover** [comment]
@jaclyn-taroni I don't quite understand. Do you mean the empty file is caused by filtering PASS?

**Comment by jaclyn-taroni** [comment]
Ah, okay - in https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/449#issuecomment-580388297 I was asking to make sure it wasn't that `01-process-sv-file.R` was doing something unintended. What

**Comment by jaclyn-taroni** [comment]
Sorry @yangyangclover - I hit enter before I finished my thought on the last comment and had to edit!

**Comment by yangyangclover** [comment]
@jaclyn-taroni  WOW! That's the reason! So what shall I do next if the CI doesn't have complete Manta data?

**Comment by jaclyn-taroni** [comment]
I am stepping out of the office for a little bit so I can’t look in depth but is there a way to skip a biospecimen if it’s not in both files in the code you are adding here?

**Comment by yangyangclover** [comment]
@jaclyn-taroni  yes we can. I know what to do next! 
If I understand you correctly,  the CI output is temporary, CI is just used to test code, but we will rerun all the codes using all complete data, right?

**Comment by jaclyn-taroni** [comment]
@yangyangclover yes, exactly right!

**Comment by yangyangclover** [comment]
Hi @jaclyn-taroni , I think I have fixed the problem, thank you for your notice.
This code is a little hard to understand, in short, 1. run shatterseek to get all original chromothripsis regions; 2. test the result and get the reasonable genome regions; 3. find out all chromothripsis-linked regions;

If you have any confusion, please let me know.  :-)

**Comment by jaclyn-taroni** [comment]
Thank you @yangyangclover! I'm going to start looking at this pull request this afternoon and will return a review by the end of the day tomorrow at the latest.

**Comment by jaclyn-taroni** [comment]
Hi @yangyangclover - I've opened https://github.com/AlexsLemonade/OpenPBTA-manuscript/issues/75 to keep track of adding the description of the modified ShatterSeek approach to the methods section of the OpenPBTA manuscript.

**Comment by yangyangclover** [comment]
@jaclyn-taroni  all right, I will

File: analyses/sv-analysis/02-shatterseek.R [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+# 6. PBTA_shatterseek_call_merge.csv
+# above three files are generated from shatterseek test
+# 7. PBTA_chromothripsis_newlink_region.csv
+# this file is about the chromothripsis region and chromothripsis-link region of each sample.
+# this file will be used in future analysis.
+
+## ===================== Load Packages =====================
+library(devtools)
+
+#withr::with_envvar(c(R_REMOTES_NO_ERRORS_FROM_WARNINGS="true"),remotes::install_github('parklab/ShatterSeek'))
```
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: > I think you are free to remove these lines now that we've gotten ShatterSeek installed on the Docker container and the ShatterSeek install step is commented out here.

OK, I removed it.

File: analyses/sv-analysis/02-shatterseek.R [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+
+## ===================== Root Directory =====================
+# Detect the ".git" folder -- this will in the project root directory.
+# Use this as the root directory to ensure proper sourcing of functions no
+# matter where this is called from
+root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
+
+
+## =====================  Create A Subdirectory to Hold All The Output Files ===================== 
+# save output files in "scratch/sv-shatterseek" and "analyses/sv-analysis/tables/sv-shatterseek"
```
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: > Is committing some of the output to the repository useful? Will the files contain values that can be used downstream in different analysis modules?

Those outputs will not be used in the following analysis, I will remove those codes related to unnecessary outputs.

File: analyses/sv-analysis/02-shatterseek.R [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+
+# save output
+write.csv(ss6,file.path(output_directory,"PBTA_shatterseek_call_minsize6.csv"),row.names = F)
+write.csv(ss3,file.path(output_directory,"PBTA_shatterseek_call_minsize3.csv"),row.names = F)
+write.csv(sv,file.path(output_directory,"PBTA_merge_sv.csv"),row.names=F)
+
+## ===================== Test and Filter The Combined Results, Then Ouput =====================
+# read files
+parkcall6 <- ss6
+parkcall3 <- ss3
```
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: OK!

File: analyses/sv-analysis/02-shatterseek.R [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+ext <- parkcall6$pval_exp_chr
+exct <- parkcall6$pval_exp_cluster
+
+# fdr test
+parkcall6$ft_fdr <- p.adjust(ft, method = "fdr")
+parkcall6$ent_fdr <- p.adjust(ent, method = "fdr")
+parkcall6$ext_fdr <- p.adjust(ext, method = "fdr")
+parkcall6$exct_fdr <- p.adjust(exct, method = "fdr")
+
+# sum all intra sv, including DEL, DUP and INV
```
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: Correct!

File: analyses/sv-analysis/02-shatterseek.R [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+# fdr test
+parkcall3$ft_fdr <- p.adjust(ft, method = "fdr")
+parkcall3$ent_fdr <- p.adjust(ent, method = "fdr")
+parkcall3$ext_fdr <- p.adjust(ext, method = "fdr")
+parkcall3$exct_fdr <- p.adjust(exct, method = "fdr")
+
+# sum all intra sv, including DEL, DUP and INV
+parkcall3$total_intra <- parkcall3$number_DEL + parkcall3$number_DUP + parkcall3$number_h2hINV + parkcall3$number_t2tINV
+
+# test: when total_intra > 2 and TRA >3 and ft_fdr > 0.2, call3 will be chrss, else call3 will be no
```
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: Right!

File: analyses/sv-analysis/02-shatterseek.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: > Because you're no longer reading this back in but are instead using the data frames that are already in the environment, could you alter the code above to assign the `shatterseek` output to `parkcall3` and `parkcall6` directly instead of replicating the object?

OK!

File: analyses/sv-analysis/02-shatterseek.R (hunk 6 above)
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: > Here `intra` refers to intrachromosomal structural variants, is that correct?

Correct!

File: analyses/sv-analysis/02-shatterseek.R (hunk 7 above)
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: > So these parameters (e.g., the `total_intra` threshold) are related to `min.size`, is that correct?

Correct!

File: analyses/sv-analysis/02-shatterseek.R [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+# combine call3 result to call6, combine call6 result to call3
+parkcall6$call3 <- parkcall3$call3
+parkcall3$call6 <- parkcall6$call6
+# keep col "chrom","start","end","sample","call3" and "call6"
+keeps <- c("chrom","start","end","sample","call3","call6")
+parkcall6 <- parkcall6[keeps]
+parkcall3 <- parkcall3[keeps]
+
+#  mergecall is a final test file
+mergecall <- rbind(parkcall6[parkcall6$call6 == "chrss"&parkcall6$call3 == "no",],parkcall3[parkcall3$call3 == "chrss",])
```
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: Yes, it is.

> Is this limiting the results to events/calls that are unique to each run of ShatterSeek with the different `min.size` parameter?

Yes, it is.

File: analyses/sv-analysis/02-shatterseek.R [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+# change colname
+colnames(mergecall) <- c("chr","start","end","sample","call3","call6")
+# save mergecall
+write.csv(mergecall,file.path(output_directory,"PBTA_shatterseek_call_merge.csv"),row.names=F)
+
+## ===================== Call Linking Groups =====================
+# read chromothripsis table
+chrss <- mergecall
```
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: > Similar to the comment above - can you assign the step where you're binding the results to `chrss`?

Sure!

File: analyses/sv-analysis/02-shatterseek.R [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+chrsslink1$chrss_chr <- gsub("25","M",chrsslink1$chrss_chr)
+chrsslink1$link_chromosome <- gsub("25","M",chrsslink1$link_chromosome)
+
+# add some column for group plotting
+chrsslink1$cluster_id <- paste(chrsslink1$sample,chrsslink1$link_chromosome,sep="_")
+chrsslink1$chrss_status <- gsub("no","link",chrsslink1$chrss_status)
+chrsslink1$circos_plot <- paste(chrsslink1$sample,".png",sep="")
+chrsslink1$cnv <- paste(chrsslink1$sample,".txt",sep="")
+
+write.csv(chrsslink1,file.path(output_directory,"PBTA_chromothripsis_newlink_region.csv"),row.names = F)
```
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: > I may have missed something, but it doesn't look like the output directory gets changed to `analyses/sv-analysis/tables/sv-shatterseek` anywhere in the script but this file is committed to the repository. I would probably save _all_ of the output here to `results` folder in your analysis folder (e.g., `analyses/sv-analysis/results`) to be consistent with the rest of the project.

I won't save any unnecessary files in this script, the final chromothripsis output (it is also the only output) will be saved in `analyses/sv-analysis/results` as you recommended.

File: analyses/sv-analysis/02-shatterseek.R [hunk 11]
```diff
@@ -0,0 +1,400 @@
...
+# sum all intra sv, including DEL, DUP and INV
+parkcall3$total_intra <- parkcall3$number_DEL + parkcall3$number_DUP + parkcall3$number_h2hINV + parkcall3$number_t2tINV
+
+# test: when total_intra > 2 and TRA >3 and ft_fdr > 0.2, call3 will be chrss, else call3 will be no
+parkcall3$call3 <- ifelse(((parkcall3$total_intra > 2)&(parkcall3$number_TRA > 3)&(parkcall3$ft_fdr > 0.2)),"chrss","no")
+# save parkcall3
+write.csv(parkcall3,file.path(output_directory,"PBTA_shatterseek_call_minsize3_test.csv"),row.names = F)
+
+# combine call3 result to call6, combine call6 result to call3
+parkcall6$call3 <- parkcall3$call3
```
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: Thank you for telling me this useful method！I will improve these steps！

[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: > If I understand correctly, you are joining or merging these calls on the combination of chromosome and sample identifier. One way to approach this that could guard against a situation where the samples were not in the same order would be to use [a join from `dplyr`](https://dplyr.tidyverse.org/reference/join.html) - so something like:
> 
> ```r
> parkcall6 <- parkcall3 %>%
>   # pick the relevant columns from parkcall3 to join to parkcall6
>   select(chrom, `sample`, call3) %>%
>   # join by chromosome and sample
>   inner_join(by = c("chrom", "sample"))
> ```
> 
> This code assumes you've run `library(dplyr)`.

Thank you for telling me this useful method！I will improve these steps！

File: analyses/sv-analysis/02-shatterseek.R [hunk 12]
```diff
@@ -0,0 +1,400 @@
...
+    )
+  CN_data <-
+    CNVsegs(
+      chrom = as.character(cnv_shatterseek$chrom),
+      start = cnv_shatterseek$loc.start,
+      end = cnv_shatterseek$loc.end,
+      total_cn = cnv_shatterseek$copy.num
+    )
+  
+  # run shatterseek in min.size = 6
```
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: > Also I don't see `min.Size` in the ShatterSeek documentation anywhere

you can find `min.Size` here:  [shatterseek.R](https://github.com/parklab/ShatterSeek/blob/master/R/shatterseek.R)
The shatterseek manuscript (Supplement) discussed the parameter selection, our parameters are based on that manuscript.

File: analyses/sv-analysis/02-shatterseek.R [hunk 13]
```diff
@@ -0,0 +1,400 @@
...
+exct <- parkcall3$pval_exp_cluster
+
+# fdr test
+parkcall3$ft_fdr <- p.adjust(ft, method = "fdr")
+parkcall3$ent_fdr <- p.adjust(ent, method = "fdr")
+parkcall3$ext_fdr <- p.adjust(ext, method = "fdr")
+parkcall3$exct_fdr <- p.adjust(exct, method = "fdr")
+
+# sum all intra sv, including DEL, DUP and INV
+parkcall3$total_intra <- parkcall3$number_DEL + parkcall3$number_DUP + parkcall3$number_h2hINV + parkcall3$number_t2tINV
```
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: > These steps are repeated for the two `min.Size` results - you could wrap them up into a function so you don't need to repeat the code twice. I'm including an example below that uses [`dplyr::mutate`](https://dplyr.tidyverse.org/reference/mutate.html):
> 
> ```r
> add_shatterseek_fdr <- function(shatterseek_result) {
>   shatterseek_result <- shatterseek_result %>%
>     dplyr::mutate(
>       ft_fdr = p.adjust(pval_fragment_joins, method = "fdr"),
>       ent_fdr = p.adjust(chr_breakpoint_enrichment, method = "fdr"),
>       ext_fdr = p.adjust(pval_exp_chr, method = "fdr"),
>       exct_fdr = p.adjust(pval_exp_cluster, method = "fdr"),
>       total_intra = (number_DEL + number_DUP + number_h2hINV + number_t2tINV)
>     )
>   return(shatterseek_result)
> }
> ```
> 
> This gets around assigning the columns that contain the p-values you want to adjust to their own values.

got it!

File: analyses/sv-analysis/02-shatterseek.R [hunk 14]
```diff
@@ -0,0 +1,400 @@
...
+svlist$end2 <- svlist$pos2
+svlist$size <- svlist$end2-svlist$end1
+
+# if linked chr is chrss, allow expand to 1sd
+# ss is a data.frame used to save chrss-link region and their chrss_chr
+ss <- data.frame()
+format_id <- chrss$format_id
+
+# chrsslink function
+chrsslink <- function(chrss,formatid=format_id){
```
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: OK , I will, thanks your suggestion!

[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: > Can you add comments here to indicate the purpose of the function, what the arguments are (and what format they should be in), and what it returns please?

OK, I will, thanks for your suggestion!

File: analyses/sv-analysis/02-shatterseek.R [hunk 15]
```diff
@@ -0,0 +1,400 @@
...
+round <- 0
+while (!identical(chrsslink1,chrsslink2)){
+  chrsslink1 <- chrsslink(chrsslink2)
+  chrsslink2 <- chrsslink(chrsslink1)
+  chrsslink1$chrss_chr <- sapply(strsplit(chrsslink1$chrss_chr, "_", fixed = TRUE), function(x)
+    paste(unique(x), collapse = "_"))
+  chrsslink2$chrss_chr <- sapply(strsplit(chrsslink2$chrss_chr, "_", fixed = TRUE), function(x)
+    paste(unique(x), collapse = "_"))
+  round <- round+1
+  print(round)
```
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: > might be a bit more informative

updated!

File: analyses/sv-analysis/02-shatterseek.R [hunk 16]
```diff
@@ -0,0 +1,400 @@
...
+  }
+  return(linkss3)
+}
+
+
+# link region analysis
+chrsslink1 <- chrsslink(chrss)
+chrsslink2 <- chrsslink(chrsslink1)
+
+# selfrenew until identical
```
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: > Can you add a comment explaining why this is necessary? It is hard to follow from the code alone.

OK

File: analyses/sv-analysis/02-shatterseek.R [hunk 17]
```diff
@@ -0,0 +1,400 @@
...
+while (!identical(chrsslink1,chrsslink2)){
+  chrsslink1 <- chrsslink(chrsslink2)
+  chrsslink2 <- chrsslink(chrsslink1)
+  chrsslink1$chrss_chr <- sapply(strsplit(chrsslink1$chrss_chr, "_", fixed = TRUE), function(x)
+    paste(unique(x), collapse = "_"))
+  chrsslink2$chrss_chr <- sapply(strsplit(chrsslink2$chrss_chr, "_", fixed = TRUE), function(x)
+    paste(unique(x), collapse = "_"))
+  round <- round+1
+  print(round)
+  print(nrow(chrsslink1)-nrow(chrsslink2))
```
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: > Same comment here - is there additional text you can add here so the message is more informative?

Commented out!

File: analyses/sv-analysis/README.md [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+2. chromothripsis analysis
+3. decipher signatures of structural variations
+4. to understand the potential mechanism behind tumors through structural variations analysis
+5. association of chromothripsis with other somatic alterations (i.e. TP53 status)
+6. multivariate [survival] analyses including clinical variables as well as ... chromothripsis.
+
+#### Inputs from data download
+* independent-specimens.wgs.primary-plus.tsv
+* pbta-sv-manta.tsv
+* pbta-cnv-cnvkit.seg
```
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/README.md]: > But this SEG file will eventually be included in the data download.

Updated!

File: analyses/sv-analysis/README.md [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+
+#### Order of scripts in analysis
+`01-process-sv-file.R` : This script is for generating sv file in shatterseek-read/signature-read input format
+
+`02-shatterseek.R` : This script is for analyzing chromothripsis, using modified shaterseek code
```
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/README.md]: > 
Thank you for correcting.

File: analyses/sv-analysis/README.md [hunk 20]
```diff
@@ -0,0 +1,22 @@
...
+**Module authors :** Yang Yang ([@yangyangclover](https://github.com/yangyangclover))
+
+
+This analysis is designed to 
+1. structural variations analysis
+2. chromothripsis analysis
+3. decipher signatures of structural variations
+4. to understand the potential mechanism behind tumors through structural variations analysis
+5. association of chromothripsis with other somatic alterations (i.e. TP53 status)
+6. multivariate [survival] analyses including clinical variables as well as ... chromothripsis.
```
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/README.md]: > I might take out the last two steps unless you have plans to perform the analyses in this module. I can imagine a situation where someone takes the output of this module and uses it for the last two steps in another part of the repository.

OK, I can do those studies (5 and 6) in a new repository if needed.

File: analyses/sv-analysis/02-shatterseek.R [hunk 21]
```diff
@@ -0,0 +1,400 @@
...
+    link_tra = link_tra=as.character(sort(as.numeric(unique(unlist(c(chrssid$chrss_chr[unlist(lapply(chrssid$chrss_chr, function(x,y=link_chrss) length(intersect(y, x))>0))]))))))
+    round <- round+1
+    print(round)}
+  link_chr <- paste(link_chrss,collapse="_")
+  chrsslink1$link_chromosome[j] <- link_chr
+}
+chrsslink1$chrss_chr <- sapply(chrsslink1$chrss_chr, function(x)
+  paste(unique(x), collapse = "_"))
+chrsslink1$chr <- gsub("23","X",chrsslink1$chr)
+chrsslink1$chr <- gsub("24","Y",chrsslink1$chr)
```
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: oh, actually there only need a step for `X`, because of ShatterSeek only accept `1-22` and `X`, so `Y` and  `M` won't appear in column `chr`.
I will remove `Y`.

File: analyses/sv-analysis/02-shatterseek.R [hunk 22]
```diff
@@ -0,0 +1,400 @@
...
+      # if the chromosome is not an original chrss
+      if(!(id %in% formatid)){
+        # caculate the region for the link region
+        linkss$chr <- unique(ss_id$chr)
+        linkss$start <- min(ss_id$start,ss_id$end)
+        linkss$end <- max(ss_id$start,ss_id$end)
+        linkss$chrss_chr <- paste(c(chrss[chrss$format_id==id,]$chrss_chr,paste(c(ss_id$chrss_chr),collapse="_",sep="_")),collapse="_",sep="_")
+        linkss$sample <- unique(ss_id$sample)
+        linkss$format_id <- unique(ss_id$format_id)
+        linkss$chrss_status <- "no"
```
[the remark this answers is not in the document]
Comment (yangyangclover) [reply analyses/sv-analysis/02-shatterseek.R]: > Is the only difference between if something is in the original `chrss` or is not in the original `chrss` the value for this column? It looks like the other operations are the same but I may have missed something.

Right
Assume that we find a link region (eg. BS_1234567_12), but this region is an original chromothripsis region computed by ShatterSeek, we still call it "chrss", which stands for "chromothripsis".

**Comment by yangyangclover** [comment]
Hi @jaclyn-taroni,  I have fixed the code. :-)

File: analyses/sv-analysis/02-shatterseek.R [hunk 23]
```diff
@@ -0,0 +1,381 @@
...
+      exct_fdr = p.adjust(pval_exp_cluster, method = "fdr"),
+      total_intra = (number_DEL + number_DUP + number_h2hINV + number_t2tINV)
+    )
+  return(shatterseek_result)
+}
+
+# fdr test in min.size=6
+ss6 <- add_shatterseek_fdr(ss6)
+# sum all intra sv, including DEL, DUP and INV
+ss6$total_intra <- ss6$number_DEL + ss6$number_DUP + ss6$number_h2hINV + ss6$number_t2tINV
```
[the remark this answers is not in the document]
Comment (jaclyn-taroni) [reply analyses/sv-analysis/02-shatterseek.R]: Or remove the `total_intra` from the function, which is probably better given what I have named the function.
